### PR TITLE
security page: restyle tab for pass/pin

### LINKED ISSUE #468: restyle tabs for pass/pin
2019-06-11 21 42 05 sketch cloud de50b3a3dc04

### PR DESCRIPTION
**Pull request by macsleven**
closes #468 

security page

### CODE REVIEW

File: Decred Wallet/Extensions/UIButton.swift [hunk 1]
```diff
@@ -16,14 +16,13 @@ extension UIButton {
...
-    
-    func addBorders(atPositions borderPositions: [BorderPosition], colorHex: String = "#333333", thickness: CGFloat = 1.7) {
+    func addBorders(atPositions borderPositions: [BorderPosition], colorHex: String = "#4e5f70", thickness: CGFloat = 1.7) {
         borderPositions.forEach({ borderPosition in
             self.addBorder(atPosition: borderPosition, colorHex: colorHex, thickness: thickness)
         })
     }
     
-    func addBorder(atPosition borderPosition: BorderPosition, colorHex: String = "#333333", thickness: CGFloat = 1.7) {
+    func addBorder(atPosition borderPosition: BorderPosition, colorHex: String = "#4e5f70", thickness: CGFloat = 1.7) {
```
Comment (neddoz) [inline Decred Wallet/Extensions/UIButton.swift]: For consistency could we inject in the Color object that we can can define in our `UICOLOR extensions` file instead of passing in the colorHex.

File: Decred Wallet/Features/Security/SecurityViewController.swift [hunk 2]
```diff
@@ -67,22 +69,22 @@ class SecurityViewController: SecurityBaseViewController {
     }
     
     func activatePasswordTab() {
-        tabController?.selectedIndex = 0
+        self.tabController?.selectedIndex = 0
```
Comment (neddoz) [inline Decred Wallet/Features/Security/SecurityViewController.swift]: We could use an enum of rawValue `Int` for the tabBarController Indexes. As such it would be more readable for instance, `tabController?.selectedIndex = enum.rawValue`.

Comment (itswisdomagain) [reply Decred Wallet/Features/Security/SecurityViewController.swift]: There are just 2 tabs, indices 0 and 1. We can _probably_ do without enums.

File: Decred Wallet/Extensions/UIButton.swift [hunk 3]
```diff
@@ -54,4 +53,22 @@ extension UIButton {
             self.setNeedsLayout()
         }
     }
+
+    private func imageWithColor(color: UIColor) -> UIImage? {
```
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIButton.swift]: - Is this `imageWithColor` function in any way related to `UIButton`?
- The function name isn't really quite clear to me. What does `imageWithColor` do?

File: Decred Wallet/Extensions/UIColor.swift [hunk 4]
```diff
@@ -23,6 +23,9 @@ extension UIColor {
         static let offWhite = UIColor(hex:"#F3F5F6")
         static let lightOffWhite = UIColor(hex: "#F9FBFA")
         static let yellowWarning = UIColor.init(hex: "#FFC84E")
+        static let tabActive = UIColor.init(hex: "#4e5f70")
```
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIColor.swift]: Let's refrain from naming color constants after their usages at the moment. Use actual color names for constants e.g. `white`, `black`, `navyBlue` etc.

We should be able to use this color constant on a different object such as label and not necessarily tabs only.

Also, naming colors properly helps to prevent us from unnecessarily duplicating colors and allows us to merge/reuse very similar colors.

File: Decred Wallet/Extensions/UIButton.swift [hunk 5]
```diff
@@ -54,4 +53,22 @@ extension UIButton {
...
+        context?.fill(rect)
+        
+        let image = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        
+        return image
+    }
+    
+    func setBackgroundColor(_ color: UIColor, for state: UIControl.State) {
+        self.setBackgroundImage(imageWithColor(color: color), for: state)
```
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIButton.swift]: You should be able to set background colors for different states from IB. This necessary?

File: Decred Wallet/Features/Security/SecurityViewController.swift [hunk 6]
```diff
@@ -48,6 +48,8 @@ class SecurityViewController: SecurityBaseViewController {
     override func viewDidLoad() {
         // delay before activating initial tab to allow borders show properly
         DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+             self.btnPassword.setBackgroundColor(UIColor.appColors.lightTabInactive, for: .highlighted)
+             self.btnPin.setBackgroundColor(UIColor.appColors.lightTabInactive, for: .highlighted)
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Security/SecurityViewController.swift]: The indenting here is off.

File: Decred Wallet/Extensions/UIColor.swift [hunk 7]
```diff
@@ -23,6 +23,9 @@ extension UIColor {
         static let offWhite = UIColor(hex:"#F3F5F6")
         static let lightOffWhite = UIColor(hex: "#F9FBFA")
         static let yellowWarning = UIColor.init(hex: "#FFC84E")
+        static let thinGray = UIColor.init(hex: "#a4abb1")
+        static let darkGray = UIColor.init(hex: "#4e5f70")
+        static let transparentThinGray = UIColor.init(hex: "#a4abb1", alpha: 0.3)
```
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIColor.swift]: I see you added the following new colors: 
`thinGray`
`transparentThinGray`

Why not reuse any of the following already defined colors?  `lightGray`, `lighterGray`, `offWhite` and `lightOffWhite`.

File: Decred Wallet/Extensions/UIButton.swift (hunk 3 above)
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIButton.swift]: Please attend to this previous comment on this line. https://github.com/raedahgroup/dcrios/pull/469/files#r294531264

File: Decred Wallet/Extensions/UIButton.swift [hunk 8]
```diff
@@ -54,4 +53,22 @@ extension UIButton {
...
+        context?.setFillColor(color.cgColor)
+        context?.fill(rect)
+        
+        let image = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        
+        return image
+    }
+    
+    func setBackgroundColor(_ color: UIColor, for state: UIControl.State) {
```
Comment (itswisdomagain) [inline Decred Wallet/Extensions/UIButton.swift]: The following approach is simpler. I don't see a need to create another function that is then called from this function when that other function is not needed else where nor is this function already complicated and needs to be simplified.

Let me rephrase. Here are 2 good reasons for creating functions:
- If a particular section of code becomes complex, you create new function(s) to hold some of the complex codes and invoke those functions from the previously complex code. This `setBackgroundColor` function is not complex, don't see a need to simplify it by creating a separate function and invoking it from here.

- Another rationale for creating functions is to abstract repeated code. If a block of code appears in multiple places, it becomes wise to refactor that block into a function and invoke that function from the multiple places the block of code appears. The result is that the new function created is used/called from multiple places in code. The `imageWithColor` function isn't called from anywhere else.

Conclusion: The content of the `imageWithColor` function can be placed in this `setBackgroundColor` function directly, no need for that other function.

See this sample code [from SO](https://stackoverflow.com/a/30604658/6086027);
```
func setBackgroundColor(color: UIColor, forState: UIControl.State) {
        self.clipsToBounds = true  // add this to maintain corner radius
        UIGraphicsBeginImageContext(CGSize(width: 1, height: 1))
        if let context = UIGraphicsGetCurrentContext() {
            context.setFillColor(color.cgColor)
            context.fill(CGRect(x: 0, y: 0, width: 1, height: 1))
            let colorImage = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
            self.setBackgroundImage(colorImage, for: forState)
        }
    }
```

File: Decred Wallet/Features/Security/SecurityViewController.swift [hunk 9]
```diff
@@ -48,6 +48,8 @@ class SecurityViewController: SecurityBaseViewController {
     override func viewDidLoad() {
         // delay before activating initial tab to allow borders show properly
         DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+            self.btnPassword.setBackgroundColor(UIColor.appColors.transparentThinGray, for: .highlighted)
+            self.btnPin.setBackgroundColor(UIColor.appColors.transparentThinGray, for: .highlighted)
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Security/SecurityViewController.swift]: The 2 lines of code above (for setting the background color) do not need to be inside this `asyncAfter` block. The comment on line 49 explains that the `asyncAfter` block is only necessary to handle correct border placement on the buttons. It's very safe to set background colors on the buttons as soon as the view loads.

File: Decred Wallet/Extensions/UIColor.swift (hunk 7 above)
Comment (macsleven) [reply Decred Wallet/Extensions/UIColor.swift]: There are  not new colors but form the ones used in storyboard.
